src/controllers/engine/controllerengine.cpp

```diff
@@ -421,63 +393,76 @@ bool ControllerEngine::execute(QScriptValue functionObject,
-   Output:  true if there was an exception
-   -------- ------------------------------------------------------ */
-bool ControllerEngine::checkException() {
-    if (m_pEngine == nullptr) {
-        return false;
-    }
+// Check if a script evaluation threw an exception. If so, register that the source
+// file threw and error and show error dialog.
+//
+// Input: QJSValue returned from evaluation
+// Output: true if there was an exception, false otherwise.
+QJSValue ControllerEngine::evaluateProgram(const QString& program, const QString& fileName,
+		int lineNumber) {
+	if (m_pEngine == nullptr) {
+		throw NullEngineException();
+	}
+
+	QJSValue evaluationResult = m_pEngine->evaluate(program, fileName, lineNumber);
+	handleEvaluationException(evaluationResult);
+
+    return evaluationResult;
+}
 
-    if (m_pEngine->hasUncaughtException()) {
-        QScriptValue exception = m_pEngine->uncaughtException();
-        QString errorMessage = exception.toString();
-        QString line = QString::number(m_pEngine->uncaughtExceptionLineNumber());
-        QStringList backtrace = m_pEngine->uncaughtExceptionBacktrace();
-        QString filename = exception.property("fileName").toString();
+void ControllerEngine::handleEvaluationException(QJSValue evaluationResult) {
+	// TODO: add test for this
+	if (evaluationResult.isError()) {
+		QString errorMessage = evaluationResult.toString();
+		QString line = evaluationResult.property("lineNumber").toString();
+		QString backtrace = evaluationResult.property("stack").toString();
+		QString filename = evaluationResult.property("fileName").toString();
 
-        QStringList error;
-        error << (filename.isEmpty() ? "" : filename) << errorMessage << line;
-        m_scriptErrors.insert((filename.isEmpty() ? "passed code" : filename), error);
+		QStringList error;
+		error << (filename.isEmpty() ? "" : filename) << errorMessage << line;
+		m_scriptErrors.insert((filename.isEmpty() ? "passed code" : filename), error);
 
-        QString errorText = tr("Uncaught exception at line %1 in file %2: %3")
-                .arg(line, (filename.isEmpty() ? "" : filename), errorMessage);
+		throw EvaluationException(errorMessage, line, backtrace, filename);
+	}
+}
 
-        if (filename.isEmpty())
-            errorText = tr("Uncaught exception at line %1 in passed code: %2")
-                    .arg(line, errorMessage);
+void ControllerEngine::presentErrorDialogForEvaluationException(EvaluationException exception) {
+	QString filename = exception.filename;
+	QString errorMessage = exception.errorMessage;
+	QString line = exception.line;
+	QString backtrace = exception.backtrace;
 
-        scriptErrorDialog(ControllerDebug::enabled() ?
-                QString("%1\nBacktrace:\n%2")
-                .arg(errorText, backtrace.join("\n")) : errorText);
+	QString errorText = tr("Uncaught exception at line %1 in file %2: %3")
+			.arg(line, (filename.isEmpty() ? "" : filename), errorMessage);
```

Comment:
This is quite similar to the code in ControllerEngine::evaluateScriptFile. Maybe add a function to EvaluationException to convert it to a formatted QString?

Reply:
Nevermind, see [my comment below](https://github.com/mixxxdj/mixxx/pull/1795#issuecomment-441297711)